### DISCUSSION TITLE
Authenticate while running as a service

### DISCUSSION THREAD

**Discussion by renarios** [abraunegg/onedrive#2701] [Q&A] [answered]
I created a service 'onedrive-user@.service' according the documentation. But when I start the service I see this in the logs:
```
Configuring Global Azure AD Endpoints
Authorize this app visiting:
https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=<cencored>&scope=Files.ReadWrite%20Files.ReadWrite.all%2>
Enter the response uri: Invalid uri
Could not initialize the OneDrive API
```

In my App Manager the service is enabled. 
Also when I run onedrive on the command line, I have to authenticate every time. How can I work around that?

**Comment by abraunegg** [comment]
As per the docs you must *first* authorise the client *before* enabling a systemd service.

> Also when I run onedrive on the command line, I have to authenticate every time. How can I work around that?

That is 100% odd and sounds like you are using the wong process.

What comand are you using to do this?

**Reply by renarios** [reply]
please see my reply below

**Comment by renarios** [comment]
This is the procedure I ran:
```
my_user@cloud:~$ onedrive
Configuration file successfully loaded
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=<cencored>                                                                                                                                                           &scope=Files.ReadWrite%20Files.ReadWrite.all%20Site                                                                                                                                                             s.Read.All%20Sites.ReadWrite.All%20offline_access&response_type=code&redirect_ur                                                                                                                                                             i=https://login.microsoftonline.com/common/oauth2/nativeclient


Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=<cencored>
Application has been successfully authorised, however no additional command switches were provided.

Please use --help for further assistance in regards to running this application.


my_user@cloud:~$ systemctl start 'onedrive-my_user@my_user.service'
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ===
Authentication is required to start 'onedrive-my_user@my_user.service'.
Authenticating as: my_user,,, (my_user)
Password: Failed to start onedrive-my_user@my_user.service: Connection timed out
See system logs and 'systemctl status onedrive-my_user@my_user.service' for details.
polkit-agent-helper-1: pam_authenticate failed: Authentication failure
```

**Comment by abraunegg** [comment]
@renarios 
Firstly, what application version are you using? What is the output of `onedrive --version` ?

#### Authentication
>```
>my_user@cloud:~$ onedrive
>Configuration file successfully loaded
>Configuring Global Azure AD Endpoints
>Authorize this app visiting:
>
>https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id<cencored>&scope=Files.ReadWrite%20Files.ReadWrite.all%20Sites.Read.All%20Sites.ReadWrite.All%20offline_access&response_type=code&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient
>
>
>Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=<cencored>
>Application has been successfully authorised, however no additional command switches were provided.
>
>Please use --help for further assistance in regards to running this application.
>```

When you do this, the client will have written the authentication data to your configuration directory, typically `~/.config/onedrive/`

Is there a 'refresh_token' in the application configuration directory post authentication?

Without trying to *start* a systemd service, can you actually *run* the application successfully if a 'refresh_token' is present? Can you run `onedrive --synchronize --verbose` without any error ?

#### systemd service
>```
>my_user@cloud:~$ systemctl start 'onedrive-my_user@my_user.service'
>==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ===
>Authentication is required to start 'onedrive-my_user@my_user.service'.
>Authenticating as: my_user,,, (my_user)
>Password: Failed to start onedrive-my_user@my_user.service: Connection timed out
>See system logs and 'systemctl status onedrive-my_user@my_user.service' for details.
>polkit-agent-helper-1: pam_authenticate failed: Authentication failure
>```

You are trying to run a systemd service as a different user to that of your user that authenticated the client

Because you are *redacting* things - it is difficult to help you here ... I dont know what user you are currently logged in as, what the actual user name of 'onedrive-my_user' .. but given you are having to *authenticate* to try and start that .. you are not logged in as the user you are trying to start the systemd service as.

#### Last Comments

So this also goes back to the authentication part .... *who* are you logged in as .. are you logged in as the *same* user you want to run the systemd service as ?

If not .. then that explains *why* things are not working for you.

**Reply by renarios** [reply]
My version is onedrive v2.4.13-1build1, installed with apt-get
I reran the steps to authenticate. The refresh_token was overwritten and now I can run onedrive manually without any errors.
I start the service with the same user as I am logged in: henk. When I start the service with ```systemctl enable 'onedrive-user@henk.service'```  I receive this error:
```==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-unit-files ===
Authentication is required to manage system service or unit files.
Authenticating as: henk,,, (henk)
Password: Failed to enable unit: Connection timed out
polkit-agent-helper-1: pam_authenticate failed: Authentication failure```
After trying a few times and entering the user password I receive ```==== AUTHENTICATION COMPLETE ===``` and the service runs succesfully. 
Is it possible to start the service without entering the password? i.e. after reboot?

**Comment by abraunegg** [comment, marked as answer]
> My version is onedrive v2.4.13-1build1, installed with apt-get

OK .. you have installed a 100% unsupported application version. Ubuntu Universe packages are obsolete and not supported. Please read the documentation on how to install the *correct* application version.

Please read the document on how to install correctly on Ubuntu.

**Reply by renarios** [reply]
Thanks. Because I want to run it for multiple accounts (and want the supported version, I'll try the dockers)
Thanks for now.

**Reply by abraunegg** [reply]
Do not run the Ubuntu provided version of Docker either. Install Docker the *correct* way.
